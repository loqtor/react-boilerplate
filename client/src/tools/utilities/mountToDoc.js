/*
 * Utility to provide aXe with a fake DOM to allow for accessibility testing
 * on mounted components within our Jest tests
*/
import { mountWithIntl } from './enzymeIntlTestHelper';

let wrapper;

export default function mountToDoc(component) {
  const container = mountWithIntl(component);

  if (!wrapper) {
    wrapper = document.createElement('main');
    document.body.appendChild(wrapper);
  }

  wrapper.innerHTML = '';
  wrapper.appendChild(container.getDOMNode());

  return container;
}
