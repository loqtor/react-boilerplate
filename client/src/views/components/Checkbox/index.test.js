import React from 'react';
import { mount } from 'enzyme';
import axe from 'axe-core';
import toJson from 'enzyme-to-json';

import mountToDoc from '../../../tools/utilities/mountToDoc';

import { Checkbox } from './index';

const setup = (render, props) => {
  const onChangeSpy = jest.fn();

  const defaultProps = {
    id: 'Subscribe',
    label: 'Subscribe to newsletter',
    onChange: onChangeSpy,
    name: 'Subscribe',
    value: 'Subscribe',
    className: 'Checkbox',
  };

  const component = render(<Checkbox {...defaultProps} {...props} />);

  return {
    actual: component,
    onChangeSpy,
  };
};

describe('<Checkbox />', () => {
  it('renders correctly', () => {
    const { actual } = setup(mount, {});

    expect(toJson(actual)).toMatchSnapshot();
  });

  it('renders correctly when an error is present', () => {
    const { actual } = setup(mount, { error: 'Required field' });

    expect(toJson(actual)).toMatchSnapshot();
    expect(actual.find('.Checkbox-error').text()).toBe('Required field');
  });

  it('renders correctly when is disabled', () => {
    const { actual } = setup(mount, { isDisabled: true });

    expect(toJson(actual)).toMatchSnapshot();
  });

  it('renders correctly when is isReadOnly', () => {
    const { actual } = setup(mount, { isReadOnly: true });

    expect(toJson(actual)).toMatchSnapshot();
  });

  it('calls onChangeSpy correctly', () => {
    const { actual, onChangeSpy } = setup(mount, {});

    actual.find('input').simulate('change');
    expect(onChangeSpy).toHaveBeenCalledTimes(1);
  });

  it('has no accessibility violations', (done) => {
    const { actual } = setup(mountToDoc, {});
    const componentNode = actual.getDOMNode();

    axe.run(componentNode, (err, { violations }) => {
      expect(err).toBe(null);
      expect(violations).toHaveLength(0);
      done();
    });
  });
});
