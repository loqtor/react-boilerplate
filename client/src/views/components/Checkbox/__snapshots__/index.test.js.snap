// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> renders correctly 1`] = `
<Checkbox
  className="Checkbox"
  error=""
  id="Subscribe"
  isChecked={null}
  isDisabled={false}
  isOptional={false}
  isRTLLabel={false}
  isReadOnly={false}
  isToggle={false}
  label="Subscribe to newsletter"
  name="Subscribe"
  onChange={[MockFunction]}
  value="Subscribe"
>
  <div
    className="Checkbox  "
  >
    <label
      htmlFor="Subscribe"
    >
      <input
        checked={null}
        className="Checkbox-input"
        disabled={false}
        id="Subscribe"
        name="Subscribe"
        onChange={[Function]}
        readOnly={false}
        type="checkbox"
        value="Subscribe"
      />
      <span
        className="Checkbox-label"
      >
        <span>
          Subscribe to newsletter
        </span>
      </span>
    </label>
  </div>
</Checkbox>
`;

exports[`<Checkbox /> renders correctly when an error is present 1`] = `
<Checkbox
  className="Checkbox"
  error="Required field"
  id="Subscribe"
  isChecked={null}
  isDisabled={false}
  isOptional={false}
  isRTLLabel={false}
  isReadOnly={false}
  isToggle={false}
  label="Subscribe to newsletter"
  name="Subscribe"
  onChange={[MockFunction]}
  value="Subscribe"
>
  <div
    className="Checkbox has-error "
  >
    <label
      htmlFor="Subscribe"
    >
      <input
        checked={null}
        className="Checkbox-input"
        disabled={false}
        id="Subscribe"
        name="Subscribe"
        onChange={[Function]}
        readOnly={false}
        type="checkbox"
        value="Subscribe"
      />
      <span
        className="Checkbox-label"
      >
        <span>
          Subscribe to newsletter
          <div
            className="Checkbox-error"
          >
            <p>
              Required field
            </p>
          </div>
        </span>
      </span>
    </label>
  </div>
</Checkbox>
`;

exports[`<Checkbox /> renders correctly when is disabled 1`] = `
<Checkbox
  className="Checkbox"
  error=""
  id="Subscribe"
  isChecked={null}
  isDisabled={true}
  isOptional={false}
  isRTLLabel={false}
  isReadOnly={false}
  isToggle={false}
  label="Subscribe to newsletter"
  name="Subscribe"
  onChange={[MockFunction]}
  value="Subscribe"
>
  <div
    className="Checkbox  "
  >
    <label
      htmlFor="Subscribe"
    >
      <input
        checked={null}
        className="Checkbox-input"
        disabled={true}
        id="Subscribe"
        name="Subscribe"
        onChange={[Function]}
        readOnly={false}
        type="checkbox"
        value="Subscribe"
      />
      <span
        className="Checkbox-label"
      >
        <span>
          Subscribe to newsletter
        </span>
      </span>
    </label>
  </div>
</Checkbox>
`;

exports[`<Checkbox /> renders correctly when is isReadOnly 1`] = `
<Checkbox
  className="Checkbox"
  error=""
  id="Subscribe"
  isChecked={null}
  isDisabled={false}
  isOptional={false}
  isRTLLabel={false}
  isReadOnly={true}
  isToggle={false}
  label="Subscribe to newsletter"
  name="Subscribe"
  onChange={[MockFunction]}
  value="Subscribe"
>
  <div
    className="Checkbox  "
  >
    <label
      htmlFor="Subscribe"
    >
      <input
        checked={null}
        className="Checkbox-input"
        disabled={false}
        id="Subscribe"
        name="Subscribe"
        onChange={[Function]}
        readOnly={true}
        type="checkbox"
        value="Subscribe"
      />
      <span
        className="Checkbox-label"
      >
        <span>
          Subscribe to newsletter
        </span>
      </span>
    </label>
  </div>
</Checkbox>
`;
